File: test_frontend_login.py
#!/usr/bin/env python3
"""
Test script to verify frontend login functionality through browser automation
"""

import requests
import json
import time

def test_existing_user_login():
    """Test login with the existing test user"""
    
    # Use the existing test user credentials
    test_user = {
        "email": "testuser1758805676@example.com",
        "password": "TestPassword123!"
    }
    
    print("🔐 Testing Frontend Login Flow")
    print("=" * 50)
    
    # Test backend health
    try:
        backend_response = requests.get("http://localhost:8000/health", timeout=5)
        if backend_response.status_code == 200:
            print("✅ Backend is running")
        else:
            print(f"❌ Backend health check failed: {backend_response.status_code}")
            return False
    except requests.exceptions.RequestException as e:
        print(f"❌ Backend is not accessible: {e}")
        return False
    
    # Test frontend health
    try:
        frontend_response = requests.get("http://localhost:3000", timeout=5)
        if frontend_response.status_code == 200:
            print("✅ Frontend is running")
        else:
            print(f"❌ Frontend health check failed: {frontend_response.status_code}")
            return False
    except requests.exceptions.RequestException as e:
        print(f"❌ Frontend is not accessible: {e}")
        return False
    
    # Test user login via backend API
    try:
        print(f"\n🔐 Testing login for: {test_user['email']}")
        
        login_data = {
            "email": test_user["email"],
            "password": test_user["password"]
        }
        
        response = requests.post(
            "http://localhost:8000/api/auth/login",
            json=login_data,
            headers={"Content-Type": "application/json"},
            timeout=10
        )
        
        if response.status_code == 200:
            result = response.json()
            print("✅ Login successful!")
            print(f"   User ID: {result.get('user', {}).get('id', 'N/A')}")
            print(f"   Email: {result.get('user', {}).get('email', 'N/A')}")
            print(f"   Display Name: {result.get('user', {}).get('display_name', 'N/A')}")
            
            # Check if tokens are present
            tokens = result.get('tokens', {})
            access_token = tokens.get('access_token')
            
            if access_token:
                print("✅ Access token received")
                return True, access_token, test_user
            else:
                print("❌ No access token in response")
                return False, None, None
        else:
            print(f"❌ Login failed: {response.status_code}")
            print(f"   Response: {response.text}")
            return False, None, None
            
    except requests.exceptions.RequestException as e:
        print(f"❌ Login request failed: {e}")
        return False, None, None

def test_results_page_access(access_token):
    """Test access to results page or related endpoints"""
    
    print(f"\n📊 Testing Results page access")
    
    try:
        headers = {
            "Authorization": f"Bearer {access_token}",
            "Content-Type": "application/json"
        }
        
        # Test user profile endpoint (which should be accessible)
        response = requests.get(
            "http://localhost:8000/api/auth/me",
            headers=headers,
            timeout=10
        )
        
        if response.status_code == 200:
            result = response.json()
            print("✅ User profile access successful!")
            print(f"   User: {result.get('email', 'N/A')}")
            
            # Test if we can access any upload-related endpoints
            try:
                upload_response = requests.get(
                    "http://localhost:8000/api/uploads",
                    headers=headers,
                    timeout=10
                )
                
                if upload_response.status_code in [200, 404]:  # 404 is ok if no uploads exist
                    print("✅ Upload endpoints are accessible")
                    return True
                else:
                    print(f"⚠️  Upload endpoint returned: {upload_response.status_code}")
                    return True  # Still consider it a success if user profile works
                    
            except requests.exceptions.RequestException:
                print("⚠️  Upload endpoint test failed, but user profile works")
                return True
                
        else:
            print(f"❌ User profile access failed: {response.status_code}")
            print(f"   Response: {response.text}")
            return False
            
    except requests.exceptions.RequestException as e:
        print(f"❌ Results page access test failed: {e}")
        return False

def test_frontend_routes():
    """Test that frontend routes are accessible"""
    
    print(f"\n🌐 Testing Frontend Routes")
    
    routes_to_test = [
        "/",
        "/login", 
        "/register",
        "/results"  # This might redirect if not authenticated
    ]
    
    for route in routes_to_test:
        try:
            response = requests.get(f"http://localhost:3000{route}", timeout=5)
            if response.status_code == 200:
                print(f"✅ Route {route} is accessible")
            else:
                print(f"⚠️  Route {route} returned: {response.status_code}")
        except requests.exceptions.RequestException as e:
            print(f"❌ Route {route} failed: {e}")
    
    return True

if __name__ == "__main__":
    # Test login
    login_success, access_token, test_user = test_existing_user_login()
    
    if login_success and access_token:
        # Test results page access
        results_success = test_results_page_access(access_token)
        
        # Test frontend routes
        routes_success = test_frontend_routes()
        
        if results_success:
            print("\n🎉 Login and access tests passed!")
            print("\n📋 Test Summary:")
            print("   ✅ Backend is running")
            print("   ✅ Frontend is running")
            print("   ✅ User login works")
            print("   ✅ Protected endpoints accessible")
            print("   ✅ Frontend routes accessible")
            print("\n🔑 Test User Credentials for Manual Testing:")
            print(f"   Email: {test_user['email']}")
            print(f"   Password: {test_user['password']}")
            print("\n💡 Next Steps:")
            print("   1. Open http://localhost:3000/login in browser")
            print("   2. Login with the above credentials")
            print("   3. Navigate to /results to test the page")
        else:
            print("\n❌ Results page access test failed")
    else:
        print("\n❌ Login test failed")

File: test_frontend_login.js
console.log('🔍 Testing frontend login flow...');

// Test the actual frontend login process
async function testFrontendLogin() {
  console.log('\n1. 🔐 Testing login via frontend API...');
  
  try {
    // First, let's login using the frontend's expected format
    const loginResponse = await fetch('http://localhost:8000/api/auth/login', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        email: 'testuser_1758806039@example.com',
        password: 'testpassword123'
      }),
    });

    if (!loginResponse.ok) {
      const errorData = await loginResponse.text();
      console.log('❌ Login failed:', loginResponse.status, errorData);
      return;
    }

    const loginResult = await loginResponse.json();
    console.log('✅ Login successful for:', loginResult.user.email);
    
    const accessToken = loginResult.tokens.access_token;
    
    // Now test all the endpoints that the frontend pages use
    console.log('\n2. 🧪 Testing all frontend API endpoints...');
    
    const endpoints = [
      { name: 'User Profile', url: '/user/profile', method: 'GET' },
      { name: 'User Settings', url: '/user/settings', method: 'GET' },
      { name: 'User Stats', url: '/users/stats', method: 'GET' },
      { name: 'Analytics User', url: '/analytics/user', method: 'GET' },
      { name: 'Analytics Jobs', url: '/analytics/jobs', method: 'GET' },
      { name: 'Job History', url: '/history', method: 'GET' },
      { name: 'Jobs List', url: '/jobs', method: 'GET' },
      { name: 'Auth Me', url: '/api/auth/me', method: 'GET' }
    ];
    
    const results = [];
    
    for (const endpoint of endpoints) {
      try {
        const response = await fetch(`http://localhost:8000${endpoint.url}`, {
          method: endpoint.method,
          headers: {
            'Authorization': `Bearer ${accessToken}`,
            'Content-Type': 'application/json',
          },
        });
        
        if (response.ok) {
          const data = await response.json();
          console.log(`✅ ${endpoint.name}: Success`);
          results.push({ endpoint: endpoint.name, status: 'success', code: response.status });
        } else {
          const errorText = await response.text();
          console.log(`❌ ${endpoint.name}: Failed (${response.status}) - ${errorText}`);
          results.push({ endpoint: endpoint.name, status: 'failed', code: response.status, error: errorText });
        }
      } catch (error) {
        console.log(`❌ ${endpoint.name}: Error - ${error.message}`);
        results.push({ endpoint: endpoint.name, status: 'error', error: error.message });
      }
    }
    
    console.log('\n📊 Summary of API endpoint tests:');
    const working = results.filter(r => r.status === 'success');
    const failing = results.filter(r => r.status !== 'success');
    
    console.log(`✅ Working endpoints (${working.length}):`);
    working.forEach(r => console.log(`   - ${r.endpoint}`));
    
    console.log(`❌ Failing endpoints (${failing.length}):`);
    failing.forEach(r => console.log(`   - ${r.endpoint}: ${r.code || 'Error'} - ${r.error || 'Unknown'}`));
    
    // Now let's test what happens when we simulate the frontend auth flow
    console.log('\n3. 🔄 Testing frontend authentication simulation...');
    
    // Simulate BackendAuthService behavior
    const mockBackendAuthService = {
      accessToken: accessToken,
      user: loginResult.user,
      
      isAuthenticated() {
        const result = !!this.accessToken && !!this.user;
        console.log(`   isAuthenticated(): ${result}`);
        return result;
      },
      
      async getAuthHeaders() {
        if (!this.accessToken) {
          throw new Error('No access token available');
        }
        return {
          'Authorization': `Bearer ${this.accessToken}`,
          'Content-Type': 'application/json',
        };
      }
    };
    
    // Test the API service flow
    console.log('\n4. 🌐 Testing API service authentication flow...');
    
    try {
      // Check if authenticated
      if (!mockBackendAuthService.isAuthenticated()) {
        console.log('❌ Would throw: "No active session. Please log in again."');
        return;
      }
      
      console.log('✅ User is authenticated, proceeding with API calls...');
      
      // Get auth headers
      const headers = await mockBackendAuthService.getAuthHeaders();
      console.log('✅ Auth headers obtained successfully');
      
      // Test a failing endpoint to see the exact error
      console.log('\n5. 🔍 Testing specific failing endpoint (analytics/user)...');
      
      const analyticsResponse = await fetch('http://localhost:8000/analytics/user', {
        headers: headers
      });
      
      if (analyticsResponse.ok) {
        console.log('✅ Analytics endpoint working!');
      } else {
        const errorText = await analyticsResponse.text();
        console.log(`❌ Analytics endpoint failed: ${analyticsResponse.status}`);
        console.log(`   Error details: ${errorText}`);
        
        // This is what would cause the "No active session" error in the frontend
        console.log('\n💡 This explains the "No active session" errors!');
        console.log('   The frontend interprets API failures as authentication failures');
      }
      
    } catch (error) {
      console.log(`❌ Auth flow error: ${error.message}`);
    }
    
  } catch (error) {
    console.error('❌ Test failed:', error.message);
  }
}

// Run the test
testFrontendLogin().then(() => {
  console.log('\n🎉 Frontend login test completed!');
  
  console.log('\n🔧 Recommended fixes:');
  console.log('1. Fix the failing API endpoints (404/validation errors)');
  console.log('2. Improve error handling to distinguish between auth and API errors');
  console.log('3. Add better user feedback for API endpoint issues');
  console.log('4. Consider graceful degradation for non-critical endpoints');
}).catch(console.error);